File: apps/web/lib/sync/session-ownership.test.ts
import { describe, expect, it } from "vitest";
import {
  OWNER_LEASE_TIMEOUT_MS,
  OwnershipCode,
  TRANSFER_TIMEOUT_MS,
  ackTransfer,
  authorizeWrite,
  claimStart,
  heartbeat,
  isOwnerStale,
  offerTransfer,
  originLabel,
  reclaimStale,
  winnerDeviceId
} from "./session-ownership";

function started(deviceId = "watch-1", nowEpochMs = 0) {
  const r = claimStart(null, { sessionId: "s1", deviceId, sportKey: "Run", nowEpochMs });
  if (!r.ok) throw new Error("lease must start");
  return r.lease;
}

describe("session ownership (ADR-011)", () => {
  it("blocks a second START while another device owns the session", () => {
    const watch = claimStart(null, {
      sessionId: "s1",
      deviceId: "watch-1",
      sportKey: "Run",
      nowEpochMs: 10
    });
    expect(watch.ok).toBe(true);
    const phone = claimStart(watch.ok ? watch.lease : null, {
      sessionId: "s2",
      deviceId: "phone-1",
      sportKey: "Run",
      nowEpochMs: 20
    });
    expect(phone.ok).toBe(false);
    if (!phone.ok) {
      expect(phone.code).toBe(OwnershipCode.SESSION_OWNED_BY);
      expect(phone.lease.ownerDeviceId).toBe("watch-1");
      expect(phone.lease.sessionId).toBe("s1");
    }
  });

  it("moves owner on transfer ACK without creating a second session", () => {
    const started = claimStart(null, {
      sessionId: "s1",
      deviceId: "watch-1",
      sportKey: "Run",
      nowEpochMs: 10
    });
    if (!started.ok) throw new Error("expected start");
    const offered = offerTransfer(started.lease, "web-1", 20);
    if (!offered.ok) throw new Error("expected offer");
    const acked = ackTransfer(offered.lease, "web-1", 21);
    expect(acked.ok).toBe(true);
    if (acked.ok) {
      expect(acked.lease.ownerDeviceId).toBe("web-1");
      expect(acked.lease.sessionId).toBe("s1");
      expect(acked.lease.state).toBe("ACTIVE");
    }
  });

  it("expires transfer after 8s and keeps the original owner", () => {
    const started = claimStart(null, {
      sessionId: "s1",
      deviceId: "watch-1",
      sportKey: "Run",
      nowEpochMs: 10
    });
    if (!started.ok) throw new Error("expected start");
    const offered = offerTransfer(started.lease, "web-1", 20);
    if (!offered.ok) throw new Error("expected offer");
    const late = ackTransfer(offered.lease, "web-1", 20 + TRANSFER_TIMEOUT_MS + 1);
    expect(late.ok).toBe(false);
    if (!late.ok) {
      expect(late.code).toBe(OwnershipCode.OFFER_EXPIRED);
      expect(late.lease.ownerDeviceId).toBe("watch-1");
    }
  });

  it("rejects ACK from a device that is not the offeree", () => {
    const started = claimStart(null, {
      sessionId: "s1",
      deviceId: "watch-1",
      sportKey: "Run",
      nowEpochMs: 10
    });
    if (!started.ok) throw new Error("expected start");
    const offered = offerTransfer(started.lease, "web-1", 20);
    if (!offered.ok) throw new Error("expected offer");
    const other = ackTransfer(offered.lease, "phone-1", 21);
    expect(other.ok).toBe(false);
    if (!other.ok) expect(other.code).toBe(OwnershipCode.NOT_OFFEREE);
  });

  it("maps device ids to origin chips", () => {
    expect(originLabel("SM-R860")).toBe("WATCH");
    expect(originLabel("web-1")).toBe("WEB");
    expect(originLabel("phone-1")).toBe("PHONE");
    expect(originLabel("cloud-sync")).toBe("CLOUD");
  });

  it("tie-breaks record conflicts on deviceId", () => {
    expect(winnerDeviceId(100, "aaa", 100, "zzz")).toBe("aaa");
    expect(winnerDeviceId(200, "watch", 50, "phone")).toBe("watch");
  });
});

// Estes testes espelham SessionOwnershipTest.kt. ADR-010: duas superficies podem
// MOSTRAR diferente, nao podem CALCULAR diferente. Se mexeres num lado, mexe no outro.
describe("epoch — writes from a device that lost ownership (ADR-011)", () => {
  it("bumps epoch on transfer and rejects the old owner's in-flight write", () => {
    const lease = started();
    expect(lease.epoch).toBe(0);
    const inFlightEpoch = lease.epoch;

    const offered = offerTransfer(lease, "phone-1", 1_000);
    expect(offered.ok).toBe(true);
    if (!offered.ok) return;
    const acked = ackTransfer(offered.lease, "phone-1", 1_100);
    expect(acked.ok).toBe(true);
    if (!acked.ok) return;
    expect(acked.lease.epoch).toBe(1);

    const late = authorizeWrite(acked.lease, "watch-1", inFlightEpoch);
    expect(late.ok).toBe(false);
    if (late.ok) return;
    expect(late.code).toBe(OwnershipCode.STALE_EPOCH);
  });

  it("authorizes the current owner at the current epoch", () => {
    const lease = started();
    expect(authorizeWrite(lease, "watch-1", lease.epoch).ok).toBe(true);
  });

  it("rejects a non-owner even at the current epoch", () => {
    const lease = started();
    const r = authorizeWrite(lease, "phone-1", lease.epoch);
    expect(r.ok).toBe(false);
    if (r.ok) return;
    expect(r.code).toBe(OwnershipCode.NOT_OWNER);
  });
});

describe("owner liveness (ADR-011)", () => {
  it("lets only the owner heartbeat", () => {
    const lease = started();
    expect(heartbeat(lease, "watch-1", 20_000).ok).toBe(true);
    const other = heartbeat(lease, "phone-1", 20_000);
    expect(other.ok).toBe(false);
    if (other.ok) return;
    expect(other.code).toBe(OwnershipCode.NOT_OWNER);
  });

  it("goes stale only after the lease timeout", () => {
    const lease = started();
    expect(isOwnerStale(lease, OWNER_LEASE_TIMEOUT_MS)).toBe(false);
    expect(isOwnerStale(lease, OWNER_LEASE_TIMEOUT_MS + 1)).toBe(true);
  });

  it("refuses reclaim while the owner is alive", () => {
    const lease = started();
    const r = reclaimStale(lease, "phone-1", 30_000);
    expect(r.ok).toBe(false);
    if (r.ok) return;
    expect(r.code).toBe(OwnershipCode.OWNER_ALIVE);
    expect(r.lease.ownerDeviceId).toBe("watch-1");
  });

  it("reclaims a dead owner's lease, bumps epoch, and zombies the old device", () => {
    // O relogio fica sem bateria a meio do treino.
    const lease = started();
    const now = OWNER_LEASE_TIMEOUT_MS + 1_000;
    const r = reclaimStale(lease, "phone-1", now);
    expect(r.ok).toBe(true);
    if (!r.ok) return;
    expect(r.lease.ownerDeviceId).toBe("phone-1");
    expect(r.lease.epoch).toBe(1);

    // O relogio volta a si e tenta escrever com o epoch antigo.
    const zombie = authorizeWrite(r.lease, "watch-1", 0);
    expect(zombie.ok).toBe(false);
    if (zombie.ok) return;
    expect(zombie.code).toBe(OwnershipCode.STALE_EPOCH);
  });

  it("keeps the lease alive across repeated heartbeats", () => {
    let lease = started();
    let t = 0;
    for (let i = 0; i < 10; i += 1) {
      t += 10_000;
      const r = heartbeat(lease, "watch-1", t);
      expect(r.ok).toBe(true);
      if (!r.ok) return;
      lease = r.lease;
      expect(isOwnerStale(lease, t)).toBe(false);
    }
    expect(t).toBe(100_000);
  });
});

File: android/shared/src/test/kotlin/com/fitconnect/shared/session/SessionOwnershipTest.kt
package com.fitconnect.shared.session

import com.fitconnect.shared.sync.RecordConflict
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class SessionOwnershipTest {
    @Test
    fun secondDeviceCannotStartWhileWatchOwns() {
        val watch = SessionOwnership.claimStart(
            current = null,
            sessionId = "s1",
            deviceId = "watch-1",
            sportKey = "Run",
            nowEpochMs = 10L,
        )
        assertTrue(watch is OwnershipResult.Ok)
        val lease = (watch as OwnershipResult.Ok).lease
        val phone = SessionOwnership.claimStart(
            current = lease,
            sessionId = "s2",
            deviceId = "phone-1",
            sportKey = "Run",
            nowEpochMs = 20L,
        )
        assertTrue(phone is OwnershipResult.Blocked)
        assertEquals(SessionOwnership.SESSION_OWNED_BY, (phone as OwnershipResult.Blocked).code)
        assertEquals("watch-1", phone.lease.ownerDeviceId)
    }

    @Test
    fun transferAckMovesOwnerWithoutSecondSession() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 10L) as OwnershipResult.Ok).lease
        val offered = (SessionOwnership.offerTransfer(started, "web-1", 20L) as OwnershipResult.Ok).lease
        val acked = (SessionOwnership.ackTransfer(offered, "web-1", 21L) as OwnershipResult.Ok).lease
        assertEquals("web-1", acked.ownerDeviceId)
        assertEquals("web-1", acked.session.deviceId)
        assertEquals(ActivitySessionState.ACTIVE, acked.session.state)
        assertEquals("s1", acked.session.sessionId)
    }

    @Test
    fun transferTimesOutAndOriginalOwnerKeepsSession() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 10L) as OwnershipResult.Ok).lease
        val offered = (SessionOwnership.offerTransfer(started, "web-1", 20L) as OwnershipResult.Ok).lease
        val late = SessionOwnership.ackTransfer(
            offered,
            "web-1",
            20L + SessionOwnership.TRANSFER_TIMEOUT_MS + 1L,
        )
        assertTrue(late is OwnershipResult.Blocked)
        assertEquals(SessionOwnership.OFFER_EXPIRED, (late as OwnershipResult.Blocked).code)
        assertEquals("watch-1", late.lease.ownerDeviceId)
        assertFalse(SessionOwnership.isLocked(ActivitySessionState.IDLE))
        assertTrue(SessionOwnership.isLocked(ActivitySessionState.ACTIVE))
    }

    @Test
    fun recordConflictTieBreaksOnDeviceId() {
        assertEquals(
            "aaa",
            RecordConflict.winnerDeviceId(100L, "aaa", 100L, "zzz"),
        )
        assertEquals(
            "watch",
            RecordConflict.winnerDeviceId(200L, "watch", 50L, "phone"),
        )
    }

    // ─── ADR-011: epoch ────────────────────────────────────────────────

    @Test
    fun transferBumpsEpochSoInFlightWritesFromOldOwnerAreRejected() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 10L) as OwnershipResult.Ok).lease
        assertEquals(0, started.epoch)

        // O relogio prepara uma escrita com o epoch que conhece.
        val inFlightEpoch = started.epoch

        val offered = (SessionOwnership.offerTransfer(started, "phone-1", 20L) as OwnershipResult.Ok).lease
        val acked = (SessionOwnership.ackTransfer(offered, "phone-1", 21L) as OwnershipResult.Ok).lease
        assertEquals(1, acked.epoch)

        // A escrita do relogio chega tarde: tem de ser REJEITADA, nao fundida.
        val late = SessionOwnership.authorizeWrite(acked, "watch-1", inFlightEpoch)
        assertTrue(late is OwnershipResult.Blocked)
        assertEquals(SessionOwnership.STALE_EPOCH, (late as OwnershipResult.Blocked).code)
    }

    @Test
    fun currentOwnerWriteIsAuthorized() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 10L) as OwnershipResult.Ok).lease
        val ok = SessionOwnership.authorizeWrite(started, "watch-1", started.epoch)
        assertTrue(ok is OwnershipResult.Ok)
    }

    @Test
    fun nonOwnerWriteIsRejectedEvenWithCurrentEpoch() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 10L) as OwnershipResult.Ok).lease
        val blocked = SessionOwnership.authorizeWrite(started, "phone-1", started.epoch)
        assertTrue(blocked is OwnershipResult.Blocked)
        assertEquals(SessionOwnership.NOT_OWNER, (blocked as OwnershipResult.Blocked).code)
    }

    // ─── ADR-011: liveness do dono ─────────────────────────────────────

    @Test
    fun onlyOwnerCanHeartbeat() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 10L) as OwnershipResult.Ok).lease
        val byOwner = SessionOwnership.heartbeat(started, "watch-1", 20_000L)
        assertTrue(byOwner is OwnershipResult.Ok)
        assertEquals(20_000L, (byOwner as OwnershipResult.Ok).lease.ownerHeartbeatAtEpochMs)

        val byOther = SessionOwnership.heartbeat(started, "phone-1", 20_000L)
        assertTrue(byOther is OwnershipResult.Blocked)
        assertEquals(SessionOwnership.NOT_OWNER, (byOther as OwnershipResult.Blocked).code)
    }

    @Test
    fun ownerGoesStaleOnlyAfterLeaseTimeout() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 0L) as OwnershipResult.Ok).lease
        assertFalse(SessionOwnership.isOwnerStale(started, SessionOwnership.OWNER_LEASE_TIMEOUT_MS))
        assertTrue(SessionOwnership.isOwnerStale(started, SessionOwnership.OWNER_LEASE_TIMEOUT_MS + 1L))
    }

    @Test
    fun reclaimBlockedWhileOwnerStillAlive() {
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 0L) as OwnershipResult.Ok).lease
        val tooEarly = SessionOwnership.reclaimStale(started, "phone-1", 30_000L)
        assertTrue(tooEarly is OwnershipResult.Blocked)
        assertEquals(SessionOwnership.OWNER_ALIVE, (tooEarly as OwnershipResult.Blocked).code)
        assertEquals("watch-1", tooEarly.lease.ownerDeviceId)
    }

    @Test
    fun deadOwnerLeaseIsReclaimableAndBumpsEpoch() {
        // O relogio fica sem bateria a meio do treino.
        val started = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 0L) as OwnershipResult.Ok).lease
        val now = SessionOwnership.OWNER_LEASE_TIMEOUT_MS + 1_000L
        val reclaimed = SessionOwnership.reclaimStale(started, "phone-1", now)
        assertTrue(reclaimed is OwnershipResult.Ok)
        val lease = (reclaimed as OwnershipResult.Ok).lease
        assertEquals("phone-1", lease.ownerDeviceId)
        assertEquals("phone-1", lease.session.deviceId)
        assertEquals(1, lease.epoch)
        assertEquals(ActivitySessionState.ACTIVE, lease.session.state)

        // O relogio volta a ligar-se e tenta escrever com o epoch antigo.
        val zombie = SessionOwnership.authorizeWrite(lease, "watch-1", 0)
        assertTrue(zombie is OwnershipResult.Blocked)
        assertEquals(SessionOwnership.STALE_EPOCH, (zombie as OwnershipResult.Blocked).code)
    }

    @Test
    fun heartbeatKeepsLeaseAliveAcrossIntervals() {
        var lease = (SessionOwnership.claimStart(null, "s1", "watch-1", "Run", 0L) as OwnershipResult.Ok).lease
        var t = 0L
        repeat(10) {
            t += SessionOwnership.OWNER_HEARTBEAT_INTERVAL_MS
            lease = (SessionOwnership.heartbeat(lease, "watch-1", t) as OwnershipResult.Ok).lease
            assertFalse(SessionOwnership.isOwnerStale(lease, t))
        }
        // 100 s de sessao com heartbeat: continua viva.
        assertEquals(100_000L, t)
        assertFalse(SessionOwnership.isOwnerStale(lease, t))
    }
}
